### ISSUE TITLE
Return types of concreate parent classes returning a templated type.

### ISSUE THREAD

**Issue by poolsimon** [vimeo/psalm#4406]
Can you explain why, because the type I return is of the right? When I make the class abstract and do not actually return, but return in an templated subclass it works:

Works: https://psalm.dev/r/d5ca92d610
Does not work: https://psalm.dev/r/051c4a5e38

I don't see why the second one, where I return an object does not work. This is a common pattern where the parent class returns a object, but a subclass can extend it and return a subclass of the return type of the parent class (covariance). 

If it is not right, how is the correct way to type it in this example? I can not find a way with having the property templated and pass all psalm errors. I can also not find a workaround.

**Comment by weirdan** [comment]
The second snippet is flagged because you would be able to do the following:
```php
/** @extends AbstractApp<MyConfig> */
class MyApp extends AbstractApp {}
```
Here, `MyApp::createConfig()` must return `MyConfig` (because of `@extends`). But it wouldn't, as it wasn't overridden. Basically, your default implementation directly contradicts the types you declared.

> If it is not right, how is the correct way to type it in this example?

There's no correct way to type it, the code _is_ wrong. Making that method abstract is actually a correct way to deal with this.
